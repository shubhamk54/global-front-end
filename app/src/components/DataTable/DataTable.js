import React, { Component } from 'react'
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Badge from '../Badge/Badge.js';
import './DataTable.scss';

import { abbreviateAmount } from '../../utils/dateUtils.js';

class DataTable extends Component {

    renderTableHeader(columns) {
        return columns.map((column, columnIndex) => <th key={columnIndex}>{column.title}</th>);
    }

    renderTableData(columns, data) {
        return data.map((dataRow, rowIndex) => <tr key={rowIndex}>
            {columns.map((column, columnIndex) => {
                const cellClassName = classNames(
                    'cell',
                    column.type && `cell-${column.type}`,
                );
                switch (column.type) {
                    case 'pill':
                        return <td className={cellClassName} key={columnIndex}>
                            <Badge
                                displayName={dataRow[column.dataKey].title}
                                variant={dataRow[column.dataKey].type} />
                        </td>

                    case 'amount':
                        return <td className={cellClassName}
                            key={columnIndex}>
                            {`${abbreviateAmount(dataRow[column.dataKey])} ${this.props.currency ? this.props.currency : ''}`}
                        </td>

                    default:
                        return <td className={cellClassName}
                            key={columnIndex}
                        >{dataRow[column.dataKey]}</td>
                }
            })}
        </tr>);
    }


    render() {
        return (
            this.props.data.length > 0 && <table className='data-table'>
                <tbody>
                    <tr>{this.renderTableHeader(this.props.columns)}</tr>
                    {this.renderTableData(this.props.columns, this.props.data)}
                </tbody>
            </table>
        )
    }
}



DataTable.propTypes = {
    data: PropTypes.array.isRequired,
    columns: PropTypes.array.isRequired,
    currency: PropTypes.string,
};

DataTable.defaultProps = {
    currency: 'USD',
}

export default DataTable