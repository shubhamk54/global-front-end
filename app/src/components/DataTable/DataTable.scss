.data-table {
  border-collapse: collapse;
  border: 3px solid #ddd;
  width: 100%;
  td,
  th {
    border: 1px solid #ddd;
    padding: 8px;
  }

  tr:nth-child(even) {
    background-color: #f2f2f2;
  }

  tr:hover {
    background-color: #ddd;
  }

  th {
    padding-top: 12px;
    padding-bottom: 12px;
    text-align: center;
    background-color: #2096f3;
    color: white;
  }
}
