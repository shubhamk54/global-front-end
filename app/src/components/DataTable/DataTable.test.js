import React from 'react';
import DataTable from './DataTable.js';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

const fakeProps = {
    columns: [
        {
            title: 'Id',
            dataKey: 'id',
        },
        {
            title: 'Name',
            dataKey: 'name'
        },
        {
            title: 'Start Date',
            dataKey: 'startDate',
            type: 'date',
        },
        {
            title: 'End Date',
            dataKey: 'endDate',
            type: 'date',
        },
        {
            title: 'Active',
            dataKey: 'active',
            type: 'pill',
        },
        {
            title: 'Budget',
            dataKey: 'Budget',
            type: 'amount',
        },
    ],
    data: [{ "id": 1, "name": "Divavu", "active": { title: 'Active', type: 'success' }, "startDate": "9/19/2017", "endDate": "3/9/2018", "Budget": 88377 },
    { "id": 2, "name": "Jaxspan", "active": { title: 'Inactive', type: 'danger' }, "startDate": "11/21/2017", "endDate": "2/21/2018", "Budget": 608715 },
    { "id": 3, "name": "Miboo", "active": { title: 'Active', type: 'success' }, "startDate": "11/1/2017", "endDate": "6/20/2017", "Budget": 239507 },
    { "id": 4, "name": "Trilith", "active": { title: 'Inactive', type: 'danger' }, "startDate": "8/25/2017", "endDate": "11/30/2017", "Budget": 179838 },
    { "id": 5, "name": "Layo", "active": { title: 'Active', type: 'success' }, "startDate": "11/28/2017", "endDate": "3/10/2018", "Budget": 837850 }],
    currency: 'USD',
};

describe('Snapshot testing for DataTable component', () => {

    it('matches the snapshot', () => {
        const tree = renderer.create(<DataTable {...fakeProps} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('DataTable testing render with negative data', () => {

    it('DataTable should not renders any table for empty data', () => {
        const wrapper = mount(
            <DataTable data={[]} columns={[]} />
        );
        const table = wrapper.find('.data-table');
        expect(table.length).toBe(0);
    });

    it('DataTable should render with no currency', () => {
        const wrapper = mount(
            <DataTable data={fakeProps.data} columns={fakeProps.columns} />
        );
        const tableData = wrapper.instance().renderTableData(fakeProps.columns, fakeProps.data);
        expect(tableData).toHaveLength(fakeProps.data.length);
    });

});

describe('DataTable component function testing', () => {

    let wrapper;
    beforeEach(() => {
        wrapper = mount(
            <DataTable {...fakeProps} />
        );
    });

    it('single table should render for Given column and data', () => {
        const table = wrapper.find('.data-table');
        expect(table.length).toBe(1);
    });

    it('Test for DataTable renders table column header for given list of column', () => {
        const tableColumns = wrapper.instance().renderTableHeader(fakeProps.columns);
        expect(tableColumns).toHaveLength(fakeProps.columns.length);
    });

    it('Test for DataTable renders table data for given list of data', () => {
        const tableData = wrapper.instance().renderTableData(fakeProps.columns, fakeProps.data);
        expect(tableData).toHaveLength(fakeProps.data.length);
    });

});
