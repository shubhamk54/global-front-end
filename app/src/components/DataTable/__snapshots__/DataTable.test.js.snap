// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test DataTable component matches the snapshot 1`] = `
<table
  className="data-table"
>
  <tbody>
    <tr>
      <th>
        Id
      </th>
      <th>
        Name
      </th>
      <th>
        Start Date
      </th>
      <th>
        End Date
      </th>
      <th>
        Active
      </th>
      <th>
        Budget
      </th>
    </tr>
    <tr>
      <td
        className="cell"
      >
        1
      </td>
      <td
        className="cell"
      >
        Divavu
      </td>
      <td
        className="cell cell-date"
      >
        9/19/2017
      </td>
      <td
        className="cell cell-date"
      >
        3/9/2018
      </td>
      <td
        className="cell cell-pill"
      >
        <span
          className="badge badge-success"
        >
          Active
        </span>
      </td>
      <td
        className="cell cell-amount"
      >
        88.4k USD
      </td>
    </tr>
    <tr>
      <td
        className="cell"
      >
        2
      </td>
      <td
        className="cell"
      >
        Jaxspan
      </td>
      <td
        className="cell cell-date"
      >
        11/21/2017
      </td>
      <td
        className="cell cell-date"
      >
        2/21/2018
      </td>
      <td
        className="cell cell-pill"
      >
        <span
          className="badge badge-danger"
        >
          Inactive
        </span>
      </td>
      <td
        className="cell cell-amount"
      >
        608.7k USD
      </td>
    </tr>
    <tr>
      <td
        className="cell"
      >
        3
      </td>
      <td
        className="cell"
      >
        Miboo
      </td>
      <td
        className="cell cell-date"
      >
        11/1/2017
      </td>
      <td
        className="cell cell-date"
      >
        6/20/2017
      </td>
      <td
        className="cell cell-pill"
      >
        <span
          className="badge badge-success"
        >
          Active
        </span>
      </td>
      <td
        className="cell cell-amount"
      >
        239.5k USD
      </td>
    </tr>
    <tr>
      <td
        className="cell"
      >
        4
      </td>
      <td
        className="cell"
      >
        Trilith
      </td>
      <td
        className="cell cell-date"
      >
        8/25/2017
      </td>
      <td
        className="cell cell-date"
      >
        11/30/2017
      </td>
      <td
        className="cell cell-pill"
      >
        <span
          className="badge badge-danger"
        >
          Inactive
        </span>
      </td>
      <td
        className="cell cell-amount"
      >
        179.8k USD
      </td>
    </tr>
    <tr>
      <td
        className="cell"
      >
        5
      </td>
      <td
        className="cell"
      >
        Layo
      </td>
      <td
        className="cell cell-date"
      >
        11/28/2017
      </td>
      <td
        className="cell cell-date"
      >
        3/10/2018
      </td>
      <td
        className="cell cell-pill"
      >
        <span
          className="badge badge-success"
        >
          Active
        </span>
      </td>
      <td
        className="cell cell-amount"
      >
        837.9k USD
      </td>
    </tr>
  </tbody>
</table>
`;
